mne/tests/test_evoked.py

```diff
@@ -81,7 +81,7 @@ def test_decim():
     info = create_info(n_channels, sfreq, 'eeg')
     with info._unlock():
         info['lowpass'] = sfreq_new / float(decim)
-    evoked = EvokedArray(data, info, tmin=-1)
+    evoked = EvokedArray(data, info)
```

Comment:
why change this?  I get that alignment will change now, so the exact times will change on the `assert` lines below, but an EvokedArray with negative `tmin` should still work.

Reply:
It doesn't work with the immediately following tests with a simpler `[::decim]` because that doesn't account for the API change

Reply:
I don't follow. The API change is that decim now shifts which samples it keeps in order to ensure that it keeps the sample at t=0.  That shouldn't mean that there are *no negative times*.

Reply:
If you have negative times then checking that the decimation worked properly is super complicated and done in the tests for epochs so better not to do that here.

Reply:
ah, OK, I see now. The lines 88 and 89 were hidden from me in the GitHub PR review UI.  If those lines are the issue, then it ought to be possible to revert the change to this line, and edit the starting sample (e.g., the `1` in `1::decim`) in lines 88-89 right?

Reply:
It's not that simple, see the tests in `test_epochs`. You have to compensate for the shift such that 0 is in the times.

Reply:
it *is* that simple. lines 84-90:
```py
    evoked = EvokedArray(data, info, tmin=-1)  # back the way it started
    evoked_dec = evoked.copy().decimate(decim)
    evoked_dec_2 = evoked.copy().decimate(decim, offset=1)
    evoked_dec_3 = evoked.decimate(dec_1).decimate(dec_2)
    assert_array_equal(evoked_dec.data, data[:, 4::decim])    # just change the 
    assert_array_equal(evoked_dec_2.data, data[:, 5::decim])  # starting samp
    assert_array_equal(evoked_dec.data, evoked_dec_3.data)
```

Reply:
also, please stop marking conversations as "resolved" just because you disagree. Sometimes it's obvious (like "please make this an f-string") and it's fine, but in more complex cases, whether something is "resolved" is generally best determined by the person who asked the question.

Reply:
-1 from me, the numbers `4` and `5` are picked out of nowhere, it's properly tested here:

```
raw, events, picks = _get_data()
events = events[events[:, 2] == 1][:2]
raw.load_data().pick_channels([raw.ch_names[pick] for pick in picks[::30]])
raw.info.normalize_proj()
del picks
sfreq_new = raw.info['sfreq'] / decim
with raw.info._unlock():
    raw.info['lowpass'] = sfreq_new / 12.  # suppress aliasing warnings
pytest.raises(ValueError, epochs.decimate, -1)
pytest.raises(ValueError, epochs.decimate, 2, offset=-1)
pytest.raises(ValueError, epochs.decimate, 2, offset=2)
for this_offset in range(decim):
    epochs = Epochs(raw, events, event_id,
                    tmin=-this_offset / raw.info['sfreq'], tmax=tmax,
                    baseline=None)
    idx_offsets = np.arange(decim) + this_offset
    for offset, idx_offset in zip(np.arange(decim), idx_offsets):
        expected_times = epochs.times[idx_offset::decim]
        expected_data = epochs.get_data()[:, :, idx_offset::decim]
        must_have = offset / float(epochs.info['sfreq'])
        assert (np.isclose(must_have, expected_times).any())
        ep_decim = epochs.copy().decimate(decim, offset)
        assert (np.isclose(must_have, ep_decim.times).any())
        assert_allclose(ep_decim.times, expected_times)
        assert_allclose(ep_decim.get_data(), expected_data)
        assert_equal(ep_decim.info['sfreq'], sfreq_new)

# More complicated cases
epochs = Epochs(raw, events, event_id, tmin, tmax)
expected_data = epochs.get_data()[:, :, ::decim]
expected_times = epochs.times[::decim]
for preload in (True, False):
    # at init
    epochs = Epochs(raw, events, event_id, tmin, tmax, decim=decim,
                    preload=preload)
    assert_allclose(epochs.get_data(), expected_data)
    assert_allclose(epochs.get_data(), expected_data)
    assert_equal(epochs.info['sfreq'], sfreq_new)
    assert_array_equal(epochs.times, expected_times)

    # split between init and afterward
    epochs = Epochs(raw, events, event_id, tmin, tmax, decim=dec_1,
                    preload=preload).decimate(dec_2)
    assert_allclose(epochs.get_data(), expected_data)
    assert_allclose(epochs.get_data(), expected_data)
    assert_equal(epochs.info['sfreq'], sfreq_new)
    assert_array_equal(epochs.times, expected_times)
    epochs = Epochs(raw, events, event_id, tmin, tmax, decim=dec_2,
                    preload=preload).decimate(dec_1)
    assert_allclose(epochs.get_data(), expected_data)
    assert_allclose(epochs.get_data(), expected_data)
    assert_equal(epochs.info['sfreq'], sfreq_new)
    assert_array_equal(epochs.times, expected_times)

    # split between init and afterward, with preload in between
    epochs = Epochs(raw, events, event_id, tmin, tmax, decim=dec_1,
                    preload=preload)
    epochs.load_data()
    epochs = epochs.decimate(dec_2)
    assert_allclose(epochs.get_data(), expected_data)
    assert_allclose(epochs.get_data(), expected_data)
    assert_equal(epochs.info['sfreq'], sfreq_new)
    assert_array_equal(epochs.times, expected_times)
    epochs = Epochs(raw, events, event_id, tmin, tmax, decim=dec_2,
                    preload=preload)
    epochs.load_data()
    epochs = epochs.decimate(dec_1)
    assert_allclose(epochs.get_data(), expected_data)
    assert_allclose(epochs.get_data(), expected_data)
    assert_equal(epochs.info['sfreq'], sfreq_new)
    assert_array_equal(epochs.times, expected_times)

    # decimate afterward
    epochs = Epochs(raw, events, event_id, tmin, tmax,
                    preload=preload).decimate(decim)
    assert_allclose(epochs.get_data(), expected_data)
    assert_allclose(epochs.get_data(), expected_data)
    assert_equal(epochs.info['sfreq'], sfreq_new)
    assert_array_equal(epochs.times, expected_times)

    # decimate afterward, with preload in between
    epochs = Epochs(raw, events, event_id, tmin, tmax, preload=preload)
    epochs.load_data()
    epochs.decimate(decim)
    assert_allclose(epochs.get_data(), expected_data)
    assert_allclose(epochs.get_data(), expected_data)
    assert_equal(epochs.info['sfreq'], sfreq_new)
    assert_array_equal(epochs.times, expected_times)

    # test picks when getting data
    picks = [3, 4, 7]
    d1 = epochs.get_data(picks=picks)
    d2 = epochs.get_data()[:, picks]
    assert_array_equal(d1, d2)
```

I meant that I thought I had addressed it but you're free to unresolve, sorry if that offended you.

Reply:
it's not plucked from nowhere. `4 == evoked.times.tolist().index(0) - decim` and `5` is because the second one has `offset=1`

Reply:
```py
    evoked = EvokedArray(data, info, tmin=-1)
    zero_idx = evoked.times.tolist().index(0)
    evoked_dec = evoked.copy().decimate(decim)
    evoked_dec_2 = evoked.copy().decimate(decim, offset=1)
    evoked_dec_3 = evoked.copy().decimate(dec_1).decimate(dec_2)
    start_samp = zero_idx - decim
    assert_array_equal(evoked_dec.data, data[:, start_samp::decim])
    # this has +1 because offset=1 when decimating ↓↓↓↓↓↓↓↓↓↓↓↓↓↓
    assert_array_equal(evoked_dec_2.data, data[:, (start_samp + 1)::decim])
```
you can change n_times on line 75 to whatever you like and this will still pass